src/main/java/bio/terra/common/retry/SamRetry.java

```diff
@@ -0,0 +1,188 @@
+    public interface SamFunction<R> {
+        R apply() throws ApiException, InterruptedException;
+    }
+
+
+    static <T> T retry(SamFunction<T> function)
+        throws InterruptedException {
+        SamRetry samRetry = new SamRetry();
+        return samRetry.perform(function);
+    }
+
+    // Version of the retry function for testing so that we can override timeout seconds
+    @VisibleForTesting
+    static <T> T retry(SamFunction<T> function, int timeoutSeconds)
+        throws InterruptedException {
+        SamRetry samRetry = new SamRetry(timeoutSeconds);
+        return samRetry.perform(function);
+    }
+
+    static void retry(SamVoidFunction function)
+        throws InterruptedException {
+        SamRetry samRetry = new SamRetry();
+        samRetry.performVoid(function);
+    }
+
+    // Version of the retry function for testing so that we can override timeout seconds
+    @VisibleForTesting
+    static void retry(SamVoidFunction function, int timeoutSeconds)
+        throws InterruptedException {
+        SamRetry samRetry = new SamRetry(timeoutSeconds);
+        samRetry.performVoid(function);
+    }
+
+    private <T> T perform(SamFunction<T> function) throws InterruptedException {
+        while (true) {
+            try {
+                return function.apply();
+            } catch (ApiException ex) {
+                handleApiException(ex);
+            } catch (Exception ex) {
+                throw new InternalServerErrorException("Unexpected exception type: " + ex.toString(), ex);
+            }
+            sleepBeforeRetrying();
+        }
+    }
+
+    private void performVoid(SamVoidFunction function) throws InterruptedException {
+        while (true) {
+            try {
+                function.apply();
+                return;
+            } catch (ApiException ex) {
+                handleApiException(ex);
+            } catch (Exception ex) {
+                throw new InternalServerErrorException("Unexpected exception type: " + ex.toString(), ex);
+            }
+            sleepBeforeRetrying();
+        }
+    }
+
+    private void handleApiException(ApiException ex) {
+        ErrorReportException rex = convertSAMExToCommonEx(ex);
+        if (!(rex instanceof InternalServerErrorException)) {
+            throw rex;
+        }
+        logger.info("SamRetry: caught retry-able exception: " + ex);
+        // if our operation timeout will happen before we wake up from our next retry
+        // sleep, then we give up and re-throw.
+        if (operationTimeout.minusSeconds(retrySeconds).isBefore(now())) {
+            logger.error("SamRetry: operation timed out after " + operationTimeout.toString());
+            throw rex;
+        }
+    }
+
+    private void sleepBeforeRetrying() throws InterruptedException {
+        logger.info("SamRetry: sleeping " + retrySeconds + " seconds");
+        TimeUnit.SECONDS.sleep(retrySeconds);
+
+        retrySeconds = retrySeconds + retrySeconds;
```

Comment:
isn't this exponential backoff, double everytime? javadoc on class says linear.

Reply:
Good catch. Copied both from the TDR codebase but should have caught. @ddietterich I changed the logic here to be linear.

Reply:
Our goal in TDR was to ride through long Sam outages caused by their merges into dev that could last for minutes. So exponential made sense. (We did several things to get our tests to run in the face of Sam merges...)

Linear retry sounds fine to me. If we need different behavior we can add it later when we know why we need it.

Reply:
Not trying to advocate for linear or exponential, just noting that the javadoc on this class promises linear but the impl looks exponential. Happy to go with whatever's been found in practice to be useful.